### PR TITLE
Allow to get UIElements and render them individually

### PR DESCRIPTION
**Pull request by Roshyo**
This PR adds 1 new Twig FIlter to get all blocks of a RichEditorField
and 2 functions to render elements either from an array of elements, or one by one.

This allows to render any element with a better customization (like allow to surround with custom DOM elements)

### CODE REVIEW

File: src/Twig/RichEditorExtension.php [hunk 1]
```diff
@@ -90,6 +87,25 @@ public function renderRichEditorField(string $content)
...
+     * @throws RuntimeError
+     * @throws SyntaxError
+     *
+     * @return array
+     */
+    public function getRichEditorFieldElements(string $content): array
+    {
+        $elements = json_decode($content, true);
+        if (!\is_array($elements)) {
+            return [$elements];
```
Comment (jacquesbh) [inline src/Twig/RichEditorExtension.php:103]: I'm not a huge fan of this.
In that case you could return `['foo']` (array of text elements, one actually) instead of an array of Ui Elements.

File: src/Twig/RichEditorExtension.php [hunk 2]
```diff
@@ -53,21 +47,24 @@ public function __construct(
...
      * @return array|TwigFunction[]
      */
-    public function getFunctions()
+    public function getFunctions(): array
     {
         return [
             new TwigFunction('monsieurbiz_richeditor_list_elements', [$this, 'listUiElements'], ['is_safe' => ['html', 'js']]),
             new TwigFunction('monsieurbiz_richeditor_youtube_link', [$this, 'convertYoutubeEmbeddedLink'], ['is_safe' => ['html', 'js']]),
+            new TwigFunction('monsieurbiz_richeditor_render_elements', [$this, 'renderElements'], ['is_safe' => ['html']]),
+            new TwigFunction('monsieurbiz_richeditor_render_element', [$this, 'renderElement'], ['is_safe' => ['html']]),
```
Comment (jacquesbh) [inline src/Twig/RichEditorExtension.php:67]: For me we should have one method only. No matter if we have one or more elements.
Dealing with plurality should not be done in twig.

**Review by jacquesbh — CHANGES_REQUESTED** [review]
Something good is coming!

File: README.md [hunk 3]
```diff
@@ -82,6 +82,28 @@ To display the content of the rich editor field you must call the twig filter:
...
+{% set elements = content|monsieurbiz_richeditor_get_elements %}
+```
+
+And then you can either render them all : 
+
+```twig
+{{ monsieurbiz_richeditor_render_elements(elements) }}
+```
+
+Or one buy one :
```
Comment (maximehuran) [inline README.md:100]:
```suggestion
Or one by one :
```

File: README.md [hunk 4]
```diff
@@ -82,6 +82,28 @@ To display the content of the rich editor field you must call the twig filter:
 
 You can see an example in the [test application](/tests/Application/templates/bundles/SyliusShopBundle/Product/Show/Tabs/Details/_description.html.twig)
 
+### Custom rendering of elements
+
+You can also render your elements with some custom DOM around that. To do so, you have access to a twig filter that 
+gives you the elements list :
+
+```twig
+{% set elements = content|monsieurbiz_richeditor_get_elements %}
```
Comment (maximehuran) [inline README.md:91]: Why you made a filter instead of a function ?

File: README.md [hunk 5]
```diff
@@ -82,6 +82,28 @@ To display the content of the rich editor field you must call the twig filter:
...
+```
+
+And then you can either render them all : 
+
+```twig
+{{ monsieurbiz_richeditor_render_elements(elements) }}
```
Comment (maximehuran) [inline README.md:97]: Why you made a function instead of a filter ?

File: README.md [hunk 6]
```diff
@@ -82,6 +82,28 @@ To display the content of the rich editor field you must call the twig filter:
...
+```
+
+Or one buy one :
+```twig
+{% for element in elements %}
+    {{ monsieurbiz_richeditor_render_element(element) }}
```
Comment (maximehuran) [inline README.md:103]: Why you made a function instead of a filter ?

File: src/Twig/RichEditorExtension.php [hunk 7]
```diff
@@ -99,7 +115,7 @@ public function renderRichEditorField(string $content)
      *
      * @return string
      */
-    private function renderElements(array $elements): string
+    public function renderElements(array $elements): string
```
Comment (maximehuran) [inline src/Twig/RichEditorExtension.php:118]: Why it becomes public ?

File: src/Twig/RichEditorExtension.php [hunk 8]
```diff
@@ -123,7 +139,7 @@ private function renderElements(array $elements): string
      *
      * @return string
      */
-    private function renderElement(array $element): string
+    public function renderElement(array $element): string
```
Comment (maximehuran) [inline src/Twig/RichEditorExtension.php:142]: Why it becomes public ?

**Review by maximehuran — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (Roshyo) [reply README.md]: Hahaha, thanks

File: README.md (hunk 4 above)
Comment (Roshyo) [reply README.md:91]: I can change that, indeed a function makes more sense here

File: README.md (hunk 6 above)
Comment (Roshyo) [reply README.md:103]: For a simple reason : the filter name was already taken by already existing behavior, and I couldn't think about another clear name. Do you have any in mind ?

File: README.md (hunk 5 above)
Comment (Roshyo) [reply README.md:97]: To keep consistency with the next one.

File: src/Twig/RichEditorExtension.php (hunk 7 above)
Comment (Roshyo) [reply src/Twig/RichEditorExtension.php:118]: To be able to use them from filter/function

File: src/Twig/RichEditorExtension.php (hunk 8 above)
Comment (Roshyo) [reply src/Twig/RichEditorExtension.php:142]: To be able to use them from filter/function

**Comment by maximehuran** [comment]
Maybe the current `monsieurbiz_richeditor_render_element` filter have to be renamed `monsieurbiz_richeditor_render_field` ?
This will let you use filters for `monsieurbiz_richeditor_render_elements` and `monsieurbiz_richeditor_render_element` as filter. We also have to see how to avoid the use of two filters for the plurality.

You can use function for `monsieurbiz_richeditor_get_elements`, it makes more sense.

File: README.md (hunk 6 above)
Comment (maximehuran) [reply README.md:103]: Maybe the current `monsieurbiz_richeditor_render_element` filter have to be renamed `monsieurbiz_richeditor_render_field` ?

Comment (Roshyo) [reply README.md:103]: Ok, I did not want to introduce any BC break. But that makes perfect sense

Comment (maximehuran) [reply README.md:103]: @jacquesbh WDYTOT ?

Comment (jacquesbh) [reply README.md:103]: Erf. I agree with both of you. The name is not clear and a BC break is not ideal.

Let's do this since we are in RC, it's time to change it, otherwise it'll be too late.

Comment (jacquesbh) [reply README.md:103]: Maybe we can make a dedicated PR for this to give us some hours to update other plugins?

Comment (jacquesbh) [reply README.md:103]: And I realize that we need a changelog 😅

File: src/Twig/RichEditorExtension.php [hunk 9]
```diff
@@ -26,48 +26,56 @@
 
 final class RichEditorExtension extends AbstractExtension
 {
-    /**
-     * @var RegistryInterface
-     */
-    private $uiElementRegistry;
+    private RegistryInterface $uiElementRegistry;
```
Comment (maximehuran) [inline src/Twig/RichEditorExtension.php:29]: PHP Doc

File: src/Twig/RichEditorExtension.php [hunk 10]
```diff
@@ -26,48 +26,56 @@
...
 
-    /**
-     * @var Environment
-     */
-    private $twig;
+    private Environment $twig;
```
Comment (maximehuran) [inline src/Twig/RichEditorExtension.php:31]: Why did you remove PHP Doc ?

File: src/Twig/RichEditorExtension.php [hunk 11]
```diff
@@ -26,48 +26,56 @@
...
+    private string $defaultElement;
```
Comment (maximehuran) [inline src/Twig/RichEditorExtension.php:32]: PHP Doc

File: src/Twig/RichEditorExtension.php [hunk 12]
```diff
@@ -26,48 +26,56 @@
...
+
+    private string $defaultElementDataField;
```
Comment (maximehuran) [inline src/Twig/RichEditorExtension.php:34]: PHP Doc

File: src/Twig/RichEditorExtension.php [hunk 13]
```diff
@@ -90,6 +98,29 @@ public function renderRichEditorField(string $content)
...
+    /**
+     * @param string $content
+     *
+     * @throws LoaderError
+     * @throws RuntimeError
+     * @throws SyntaxError
+     *
+     * @return array
+     */
+    public function getRichEditorFieldElements(string $content): array
```
Comment (maximehuran) [inline src/Twig/RichEditorExtension.php:110]: `getRichEditorElements` seems more logic

Comment (maximehuran) [reply src/Twig/RichEditorExtension.php:110]: Or `getElements` to be simple.

File: src/Twig/RichEditorExtension.php [hunk 14]
```diff
@@ -80,7 +88,7 @@ public function getFunctions()
      *
      * @return string
      */
-    public function renderRichEditorField(string $content)
+    public function renderRichEditorField(string $content): string
```
Comment (maximehuran) [inline src/Twig/RichEditorExtension.php:91]: If we change `getRichEditorElements` to `getElements` we can change here to `renderField`

File: src/Twig/RichEditorExtension.php (hunk 10 above)
Comment (Roshyo) [reply src/Twig/RichEditorExtension.php:31]: I removed it because the PHP 7.4 strict typing is enough, isn't it ?

Comment (maximehuran) [reply src/Twig/RichEditorExtension.php:31]: My bad it's good

**Review by maximehuran — APPROVED** [review]
SGTM last thing will be to have the tests run

File: src/Twig/RichEditorExtension.php [hunk 15]
```diff
@@ -90,6 +99,29 @@ public function renderRichEditorField(string $content)
...
+     * @return array
+     */
+    public function getElements(string $content): array
+    {
+        $elements = json_decode($content, true);
+        if (!\is_array($elements)) {
+            // If the JSON decode failed, return a new UIElement with default configuration
+            return [
+                'type' => $this->getDefaultElement(),
+                'data' => $content,
```
Comment (jacquesbh) [inline src/Twig/RichEditorExtension.php:118]:
```suggestion
                'data' => [$this->getDefaultElementDataField()  => $content],
```

Comment (Roshyo) [reply src/Twig/RichEditorExtension.php]: Ha, you are absolutely right !

**Review by jacquesbh — APPROVED** [review]

**Comment by jacquesbh** [comment]
Thank you @Roshyo !! 😍

I'll release the RC.8 right now :).
